azure-iot-device/azure/iot/device/common/auth/sastoken.py

```diff
@@ -0,0 +1,96 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""This module contains tools for working with Shared Access Signature (SAS) Tokens"""
+
+import base64
+import hmac
+import hashlib
+import time
+import six.moves.urllib as urllib
+from azure.iot.device.common.chainable_exception import ChainableException
+
+
+class SasTokenError(ChainableException):
+    """Error in SasToken"""
+
+    pass
+
+
+class SasToken(object):
+    """Shared Access Signature Token used to authenticate a request
+
+    Data Attributes:
+    expiry_time (int): Time that token will expire (in UTC, since epoch)
+    ttl (int): Time to live for the token, in seconds
+    """
+
+    _auth_rule_token_format = (
+        "SharedAccessSignature sr={resource}&sig={signature}&se={expiry}&skn={keyname}"
+    )
+    _simple_token_format = "SharedAccessSignature sr={resource}&sig={signature}&se={expiry}"
+
+    def __init__(self, uri, signing_mechanism, key_name=None, ttl=3600):
+        """
+        :param str uri: URI of the resouce to be accessed
+        :param signing_mechanism: The signing mechanism to use in the SasToken
+        :type signing_mechanism: Child classes of :class:`azure.iot.common.SigningMechanism`
+        :param str key_name: Symmetric Key Name (optional)
+        :param int ttl: Time to live for the token, in seconds (default 3600)
+
+        :raises: SasTokenError if an error occurs building a SasToken
+        """
+        self._uri = uri
+        self._signing_mechanism = signing_mechanism
+        self._key_name = key_name
+        self._expiry_time = None  # This will be overwritten by the .refresh() call below
+        self._token = None  # This will be overwritten by the .refresh() call below
+
+        self.ttl = ttl
+        self.refresh()
+
+    def __str__(self):
+        return self._token
```

Comment:
Mental exercise :#2 (written after the one in pipeline_stages_base.py).  

postulate: The notion that a SasToken has a value and that we need to periodically "refresh" that value is not quite right.  It's better to think about the SasToken object as a "password generator" object.  When a consumer of this class needs to talk to a network service, he can use this object to generate a password that gives him access for some period of time.  When that password is set to expire, the consumer is responsible for generating a new one.   This is how the code is written and I like it.

Saying `token.refresh(); password=str(token);` as we do, accomplishes this, but it's not quite semantically valid().  Rather, what we want is `password=token.generate_new_password()`

why this matters:  It doesn't _really_ matter that much, but, there an odd ownership relationship with the root owning the SasToken, object and 2 different pipelines controlling it.  Both MQTT and HTTP can refresh the token.  If we think of the SasToken object as having a value, having someone else change the value feels wrong.  But if we think of each pipeline as owning it's own password that it gets from the generator object, it just feels better.